File: go/src/args.go
package main

import "flag"

type Args struct {
	DebugCpu bool
	DebugApu bool
	DebugGpu bool
	DebugRam bool
	Headless bool
	Silent   bool
	Turbo    bool
	Profile  int
	Rom      string
}

func boolFlagWithShort(long string, short string, default_value bool, help string) *bool {
	var result bool

	flag.BoolVar(&result, long, default_value, help)
	flag.BoolVar(&result, short, default_value, help)

	return &result
}

func NewArgs() *Args {
	var debug_cpu = boolFlagWithShort("debug-cpu", "c", false, "Debug CPU")
	var debug_gpu = boolFlagWithShort("debug-gpu", "g", false, "Debug GPU")
	var debug_apu = boolFlagWithShort("debug-apu", "a", false, "Debug APU")
	var debug_ram = boolFlagWithShort("debug-ram", "r", false, "Debug RAM")
	var headless = boolFlagWithShort("headless", "H", false, "No video")
	var silent = boolFlagWithShort("silent", "S", false, "No audio")
	var turbo = boolFlagWithShort("turbo", "t", false, "No sleep()")

	var profile int
	flag.IntVar(&profile, "profile", 0, "Exit after N frames")
	flag.IntVar(&profile, "p", 0, "Exit after N frames")

	flag.Parse()
	var rom = flag.Arg(0)

	return &Args{
		DebugCpu: *debug_cpu,
		DebugGpu: *debug_gpu,
		DebugApu: *debug_apu,
		DebugRam: *debug_ram,
		Headless: *headless,
		Silent:   *silent,
		Turbo:    *turbo,
		Profile:  profile,
		Rom:      rom,
	}
}

File: nim/src/args.nim
import std/os

import argparse

type
    Args* = object
        rom*: string
        frames*, profile*: int
        headless*, silent*: bool
        turbo*, debug_cpu*, debug_apu*, debug_gpu*, debug_ram*: bool

proc parseArgs*(args: seq[string]): Args =
    var p = newParser("rosettaboy-nim"):
        flag("-H", "--headless")
        flag("-S", "--silent")
        flag("-c", "--debug-cpu")
        flag("-g", "--debug-gpu")
        flag("-r", "--debug-ram")
        flag("-a", "--debug-apu")
        flag("-t", "--turbo")
        option("-f", "--frames", default = some("0"), help = "Exit after N frames")
        option("-p", "--profile", default = some("0"), help = "Exit after N seconds")
        arg("rom")

    try:
        var opts = p.parse(args)
        return Args(
            rom: opts.rom,
            headless: opts.headless,
            silent: opts.silent,
            debug_cpu: opts.debug_cpu,
            debug_gpu: opts.debug_gpu,
            debug_apu: opts.debug_apu,
            debug_ram: opts.debug_ram,
            frames: parseInt(opts.frames),
            profile: parseInt(opts.profile),
            turbo: opts.turbo,
        )
    except ShortCircuit as e:
        if e.flag == "argparse_help":
            echo p.help
            quit(0)
